src/rules/noAsyncWithoutAwaitRule.ts

```diff
@@ -0,0 +1,139 @@
+        optionExamples: [true],
+        options: null,
+        optionsDescription: "Not configurable.",
+        /* tslint:disable:max-line-length */
+        rationale: Lint.Utils.dedent`
+        Marking a function as \`async\` without using \`await\` or returning a value inside it can lead to an unintended promise return and a larger transpiled output.
+        Often the function can be synchronous and the \`async\` keyword is there by mistake.
+        Return statements are allowed as sometimes it is desirable to wrap the returned value in a Promise`,
+        /* tslint:enable:max-line-length */
+        ruleName: "no-async-without-await",
+        type: "functionality",
+        typescriptOnly: false,
+    };
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithWalker(new Walk(sourceFile, this.getOptions()));
+    }
+}
+
+class Walk extends Lint.RuleWalker {
+    protected visitFunctionDeclaration(node: ts.FunctionDeclaration) {
+        this.addFailureIfAsyncFunctionHasNoAwait(node);
+        super.visitFunctionDeclaration(node);
+    }
+
+    protected visitArrowFunction(node: ts.ArrowFunction) {
+        this.addFailureIfAsyncFunctionHasNoAwait(node);
+        super.visitArrowFunction(node);
+    }
+
+    protected visitMethodDeclaration(node: ts.MethodDeclaration) {
+        this.addFailureIfAsyncFunctionHasNoAwait(node);
+        super.visitMethodDeclaration(node);
+    }
+
+    protected visitFunctionExpression(node: ts.FunctionExpression) {
+        this.addFailureIfAsyncFunctionHasNoAwait(node);
+        super.visitFunctionExpression(node);
+    }
+
+    private getAsyncModifier(node: ts.Node) {
+        if (node.modifiers !== undefined) {
+            return node.modifiers.find((modifier) => modifier.kind === ts.SyntaxKind.AsyncKeyword);
+        }
+
+        return undefined;
+    }
+
+    private isAwait(node: ts.Node): boolean {
+        return node.kind === ts.SyntaxKind.AwaitKeyword;
+    }
+
+    private isReturn(node: ts.Node): boolean {
+        return node.kind === ts.SyntaxKind.ReturnKeyword;
+    }
+
+    private functionBlockHasAwait(node: ts.Node): boolean {
+        if (this.isAwait(node)) {
+            return true;
+        }
+
+        if (node.kind === ts.SyntaxKind.ArrowFunction
+            || node.kind === ts.SyntaxKind.FunctionDeclaration) {
+            return false;
+        }
+
+        // tslint:disable-next-line:no-unsafe-any
+        return node.getChildren().some(this.functionBlockHasAwait.bind(this));
+    }
+
+    private functionBlockHasReturn(node: ts.Node): boolean {
+        if (this.isReturn(node)) {
+            return true;
+        }
+
+        if (node.kind === ts.SyntaxKind.ArrowFunction
+            || node.kind === ts.SyntaxKind.FunctionDeclaration) {
+            return false;
+        }
+
+        const returnInChildren = node.getChildren()
+            .map((functionBlockNode: ts.Node) => this.functionBlockHasReturn(functionBlockNode));
```

Comment:
> `.map`
> `.some`

Can you simplify this into `.some(functionBlockHasReturn)`, once it's moved to a standalone function declaration? Should be a bit simpler, right?